clients/tabby-agent/src/lsp/ChatEditProvider.ts

```diff
@@ -249,121 +256,129 @@ export class ChatEditProvider {
     responseDocumentTag: string[],
     responseCommentTag?: string[],
   ): Promise<void> {
-    const finalize = async (state: "completed" | "stopped") => {
-      if (this.currentEdit) {
-        const edit = this.currentEdit;
-        edit.state = state;
-        const editedLines = this.generateChangesPreview(edit);
-        await this.connection.workspace.applyEdit({
-          edit: {
-            changes: {
-              [edit.location.uri]: [
-                {
-                  range: edit.editedRange,
-                  newText: editedLines.join("\n") + "\n",
-                },
-              ],
+    const applyEdit = async (edit: Edit, isFirst: boolean = false, isLast: boolean = false) => {
+      const editedLines = this.generateChangesPreview(edit);
+      const workspaceEdit: WorkspaceEdit = {
+        changes: {
+          [edit.location.uri]: [
+            {
+              range: edit.editedRange,
+              newText: editedLines.join("\n") + "\n",
             },
-          },
-        });
+          ],
+        },
+      };
+
+      await this.applyWorkspaceEdit({
+        edit: workspaceEdit,
+        options: {
+          undoStopBefore: isFirst,
+          undoStopAfter: isLast,
+        },
+      });
+
+      edit.editedRange = {
+        start: { line: edit.editedRange.start.line, character: 0 },
+        end: { line: edit.editedRange.start.line + editedLines.length, character: 0 },
+      };
+    };
+
+    const processBuffer = (edit: Edit, inTag: "document" | "comment", closeTag: string) => {
+      const reg = this.createCloseTagMatcher(closeTag);
+      const match = reg.exec(edit.buffer);
+      if (!match) {
+        edit[inTag === "document" ? "editedText" : "comments"] += edit.buffer;
+        edit.buffer = "";
+      } else {
+        edit[inTag === "document" ? "editedText" : "comments"] += edit.buffer.substring(0, match.index);
+        edit.buffer = edit.buffer.substring(match.index);
+        return match[0] === closeTag ? false : inTag;
       }
-      this.currentEdit = null;
-      this.mutexAbortController = null;
+      return inTag;
     };
+
+    const findOpenTag = (
+      buffer: string,
+      responseDocumentTag: string[],
+      responseCommentTag?: string[],
+    ): "document" | "comment" | false => {
+      const openTags = [responseDocumentTag[0], responseCommentTag?.[0]].filter(Boolean);
+      if (openTags.length < 1) return false;
+
+      const reg = new RegExp(openTags.join("|"), "g");
+      const match = reg.exec(buffer);
+      if (match && match[0]) {
+        if (match[0] === responseDocumentTag[0]) {
+          return "document";
+        } else if (match[0] === responseCommentTag?.[0]) {
+          return "comment";
+        }
+      }
+      return false;
+    };
```

Comment:
It seems there is a bug after the refactor of `findOpenTag` and `processBuffer`. The opening tag (`<GENERATEDCODE>`) is not being skipped after it is found, instead, it is always included in the content after editing.

Reply:
I'm sorry, I made a logical mistake, but I have fixed it. Sorry for the late response.